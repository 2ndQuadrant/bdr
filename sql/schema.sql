--
-- BDR tables' attributes must never change with schema changes.
-- Only new attributes can be appended and only if nullable.
--

select attrelid::regclass::text, attnum, attname, attisdropped, atttypid::regtype, attnotnull
from pg_attribute
where attrelid = ANY (ARRAY['bdr.bdr_nodes', 'bdr.bdr_connections', 'bdr.bdr_queued_drops', 'bdr.bdr_queued_commands']::regclass[])
  and attnum >= 1
order by attrelid, attnum;
