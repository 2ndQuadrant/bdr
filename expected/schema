-- RT#60660
SELECT * FROM public.bdr_regress_variables()
\gset
\c :writedb1
SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE TABLE public.desync (
   id integer primary key not null,
   n1 integer not null
);
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

\d desync
    Table "public.desync"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
 n1     | integer | not null
Indexes:
    "desync_pkey" PRIMARY KEY, btree (id)

SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- basic builtin datatypes
\c :writedb2
\d desync
    Table "public.desync"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
 n1     | integer | not null
Indexes:
    "desync_pkey" PRIMARY KEY, btree (id)

\c :writedb1
-- Add a new attribute on this node only
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
COMMIT;
-- Create an natts=3 tuple with null RHS. This should apply fine on the other
-- side, it'll disregard the righthand NULL.
INSERT INTO desync(id, n1, dropme) VALUES (1, 1, NULL);
SELECT * FROM desync ORDER BY id;
 id | n1 | dropme 
----+----+--------
  1 |  1 |       
(1 row)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '2s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
-- Drop the attribute; we're still natts=3, but one is dropped
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;
-- create second natts=3 tuple on db1
--
-- This will also apply on the other side, because dropped columns are always
-- sent as nulls.
INSERT INTO desync(id, n1) VALUES (2, 2);
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
(2 rows)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '2s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
\c :writedb2
-- Both new tuples should've arrived
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
(2 rows)

-- create natts=2 tuple on db2
--
-- This should apply to writedb1 because we right-pad rows with low natts if
-- the other side col is dropped (or nullable)
INSERT INTO desync(id, n1) VALUES (3, 3);
SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

-- Make our side confirm to the remote schema again
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;
\c :writedb1
-- So now this side should apply too
SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
-- Yay!
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

\c :writedb2
-- Yay! Again!
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

SELECT bdr.bdr_replicate_ddl_command($DDL$
DROP TABLE public.desync;
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

