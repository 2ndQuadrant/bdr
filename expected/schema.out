--
-- BDR tables' attributes must never change with schema changes.
-- Only new attributes can be appended and only if nullable.
--
select attrelid::regclass::text, attnum, attname, attisdropped, atttypid::regtype, attnotnull
from pg_attribute
where attrelid = ANY (ARRAY['bdr.bdr_nodes', 'bdr.bdr_connections', 'bdr.bdr_queued_drops', 'bdr.bdr_queued_commands']::regclass[])
  and attnum >= 1
order by attrelid, attnum;
        attrelid         | attnum |        attname         | attisdropped |         atttypid         | attnotnull 
-------------------------+--------+------------------------+--------------+--------------------------+------------
 bdr.bdr_connections     |      1 | conn_sysid             | f            | text                     | t
 bdr.bdr_connections     |      2 | conn_timeline          | f            | oid                      | t
 bdr.bdr_connections     |      3 | conn_dboid             | f            | oid                      | t
 bdr.bdr_connections     |      4 | conn_origin_sysid      | f            | text                     | t
 bdr.bdr_connections     |      5 | conn_origin_timeline   | f            | oid                      | t
 bdr.bdr_connections     |      6 | conn_origin_dboid      | f            | oid                      | t
 bdr.bdr_connections     |      7 | conn_is_unidirectional | f            | boolean                  | t
 bdr.bdr_connections     |      8 | conn_dsn               | f            | text                     | t
 bdr.bdr_connections     |      9 | conn_apply_delay       | f            | integer                  | f
 bdr.bdr_connections     |     10 | conn_replication_sets  | f            | text[]                   | f
 bdr.bdr_nodes           |      1 | node_sysid             | f            | text                     | t
 bdr.bdr_nodes           |      2 | node_timeline          | f            | oid                      | t
 bdr.bdr_nodes           |      3 | node_dboid             | f            | oid                      | t
 bdr.bdr_nodes           |      4 | node_status            | f            | "char"                   | t
 bdr.bdr_nodes           |      5 | node_name              | f            | text                     | t
 bdr.bdr_nodes           |      6 | node_local_dsn         | f            | text                     | f
 bdr.bdr_nodes           |      7 | node_init_from_dsn     | f            | text                     | f
 bdr.bdr_nodes           |      8 | node_read_only         | f            | boolean                  | f
 bdr.bdr_nodes           |      9 | node_seq_id            | f            | smallint                 | f
 bdr.bdr_queued_commands |      1 | lsn                    | f            | pg_lsn                   | t
 bdr.bdr_queued_commands |      2 | queued_at              | f            | timestamp with time zone | t
 bdr.bdr_queued_commands |      3 | perpetrator            | f            | text                     | t
 bdr.bdr_queued_commands |      4 | command_tag            | f            | text                     | t
 bdr.bdr_queued_commands |      5 | command                | f            | text                     | t
 bdr.bdr_queued_commands |      6 | search_path            | f            | text                     | t
 bdr.bdr_queued_drops    |      1 | lsn                    | f            | pg_lsn                   | t
 bdr.bdr_queued_drops    |      2 | queued_at              | f            | timestamp with time zone | t
 bdr.bdr_queued_drops    |      3 | dropped_objects        | f            | bdr.dropped_object[]     | t
(28 rows)

